src/cedalion/sigproc/motion_correct.py

```diff
@@ -443,3 +447,94 @@ def motion_correct_PCA_recurse(
         tInc.values = np.hstack([tInc.values[0], tInc.values[:-1]])
 
     return fNIRSdata_cleaned, svs, nSV_ret, tInc
+
+
+def tddr(ts: cdt.NDTimeSeries):
+    """Implementation of the TDDR algorithm for motion correction.
+
+    Uses an iterative reweighting approach to reduce large fluctuations typically
+    associated with motion artifacts. Adapted for cedalion from the python
+    implementation at :cite:`Fishburn2018`, which is the reference implementation for
+    the algorithm described in :cite:`Fishburn2019`.
+
+    Arguments:
+        ts: The time series to be corrected. Should have dims channel and wavelength
+
+    Returns:
+        The corrected time series.
+
+    References:
+        Paper: :cite:`Fishburn2019`
+        Code: :cite:`Fishburn2018`
+    """
+    signal = ts.copy()
+    unit = signal.pint.units if signal.pint.units else 1
+    signal = signal.pint.dequantify()
+
+    if signal.channel.size != 1 or signal.wavelength.size != 1:
+        for ch in signal.channel.values:
+            for wl in signal.wavelength.values:
+                # Select single channel/wavelength
+                curr_signal = signal.loc[dict(channel=[ch], wavelength=[wl])]
+                # Process the single time series
+                corrected = tddr(curr_signal)
+                # Assign back ensuring coordinate consistency
+                signal.loc[dict(channel=[ch], wavelength=[wl])] = corrected
+        return signal
+
+    # Preprocess: Separate high and low frequencies
+    signal_mean = np.mean(signal)
+    signal -= signal_mean
+    signal_low = signal.cd.freq_filter(0, 0.5, 3)
+    signal_high = signal - signal_low
+
+    # Initialize
+    tune = 4.685
+    D = np.sqrt(np.finfo(signal.dtype).eps)
+    mu = np.inf
+    iter = 0
+
+    # Step 1. Compute temporal derivative of the signal
+    deriv = np.diff(signal_low)
+
+    # Step 2. Initialize observation weights
+    w = np.ones(deriv.shape)
+
+    # Step 3. Iterative estimation of robust weights
+    while iter < 50:
```

Comment:
- probably, this should be a for loop with an else clause that warns if the termination criterion was not reached
- please rename iter to n_iter